File: dynamic_programming/max_subarray.kt
fun main() {
    val s = Solution()
    val nums = intArrayOf(-2,1,-3,4,-1,2,1,-5,4)
    println(s.maxSubArray(nums))
}

class Solution {
    fun maxSubArray(nums: IntArray): Int {

        if (nums.size == 0) {
            return 0
        }

        var maxSum = nums[0]
        var curSum = maxSum

        for (i in 1 until nums.size) {
            curSum = maxOf(nums[i], curSum + nums[i])
            maxSum = maxOf(curSum, maxSum)
        }

        return maxSum
    }
}

File: dynamic_programming/max_subarray.py
#! /usr/bin/env python3

class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        if not nums:
            return 0

        curSum = nums[0]
        maxSum = curSum
        for n in nums[1:]:
            curSum = max(n, curSum + n)
            maxSum = max(maxSum, curSum)
        return maxSum


if __name__ == "__main__":
    s = Solution()
    nums = [-2,1,-3,4,-1,2,1,-5,4]
    print(s.maxSubArray(nums))
